parsing/lexer.mll

```diff
@@ -190,6 +200,29 @@ let char_for_hexadecimal_code lexbuf i =
   in
   Char.chr (val1 * 16 + val2)
 
+let uchar_for_uchar_escape lexbuf =
+  let len = Lexing.lexeme_end lexbuf - Lexing.lexeme_start lexbuf in
+  let spos = 3 (* skip opening \u{ *) in
+  let epos = len - 2 (* skip closing } *) in
+  let rec hex_num_value acc i =
+    if i > epos then acc else
+    let digit = Char.code (Lexing.lexeme_char lexbuf i) in
+    let value =
+      if digit >= 97 then digit - 87 else
+      if digit >= 65 then digit - 55 else
+      digit - 48
+    in
+    hex_num_value (16 * acc + value) (i + 1)
+  in
+  let cp = hex_num_value 0 spos in
+  if Uchar.is_valid cp then Uchar.unsafe_of_int cp else
+  raise
+    (Error
+       (Illegal_escape
+          (Lexing.lexeme lexbuf ^ ", " ^ Printf.sprintf "%X" cp ^
```

Comment:
The lexbuf content is what the user wrote which may have for example leading zeroes. The parsed `cp` what the program understood. Showing how the program interpreted your input sometimes leads to easier diagnostics.